README.md
# Calendar Carousel

A custom widget that display calendar in a form of carousel slider.

## Installation

Add `carousel_calendar: ^0.0.2` to your `pubspec.yaml` dependencies. And import it:

```dart
import 'package:carousel_calendar/carousel_calendar.dart';
```

## Features

* Display calendar in a form of carousel
* Listen to the unchanged events for the dates.


## Supported platforms

* Flutter Android
* Flutter iOS
* Flutter web
* Flutter desktop

## Usage

Simply create a `CalendarCarousel` widget, and pass the params according to your customization:

```dart
import 'flutter/material.dart';
import 'package:dp_widget/dp_widget.dart';

class SamplePage extends StatefulWidget {
  const SamplePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  State<SamplePage> createState() => _SamplePageState();
}

class _SamplePageState extends State<SamplePage> {
  final GlobalKey<ScaffoldState> _scaffoldKey = GlobalKey();
  DateTime? date;
  
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        key: _scaffoldKey,
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
        child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: CalendarCarousel(
                    monthSelectorMargin: EdgeInsets.only(bottom: 2.0),
                    weekdayMargin: EdgeInsets.symmetric(vertical: 12.0),
                    weekdayAbb: WeekdayAbbreviation.two,
                    dayCarouselHeight: 120,
                    showYearAlways: true,
                    onChanged: (nm) {
                      setState(() {
                        date = nm;
                      });
                    },
                  ),
                ),
                Expanded(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: <Widget>[
                      const Text(
                        'Selected Date',
                      ),
                      Text(
                        '${date != null ? "${date!.day}-${date!.month}-${date!.year}" : "No selection yet."}',
                        style: Theme.of(context).textTheme.headlineMedium,
                      ),
                    ],
                  ),
                ),
              ],
            ),
    ),),);
    }
}

```

## Screenshot

![simple](screenshot.png)

The screenshots above can be found at the example project.

## License

MIT


CHANGELOG.md
## 0.0.1

* initial release. A Widget that displays calendar in form of a carousel.

## 0.0.2

* Updated the widget to accommodate external date value, in order to be able to set and preserve the state externally.


test/calendar_carousel_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:calendar_carousel/calendar_carousel.dart';

void main() {}


lib/src/calendar_carousel.dart
import 'package:carousel_slider/carousel_slider.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

final DateTime _currentDate = DateTime.now();

Map<String, String> month({required Months month, int? year}) {
  Map<String, String> map = {};
  var days = 31;
  switch (month) {
    case Months.February:
      days = 28;
      break;
    case Months.April:
    case Months.June:
    case Months.September:
    case Months.November:
      days = 30;
      break;
    default:
      days = 31;
      break;
  }
  year = year ?? _currentDate.year;

  if ((month.index + 1) == _currentDate.month && year == _currentDate.year) {
    days = _currentDate.day;
  }

  List.generate(days, (index) => index + 1).forEach((int day) {
    map.addAll({
      day.toString(): getDay(DateTime(year!, month.index + 1, day).weekday)
    });
  });
  return map;
}

String getDay(int element) {
  switch (element % 7) {
    case 1:
      return "Monday";
    case 2:
      return "Tuesday";
    case 3:
      return "Wednesday";
    case 4:
      return "Thursday";
    case 5:
      return "Friday";
    case 6:
      return "Saturday";
    default:
      return "Sunday";
  }
}

enum Months {
  January,
  February,
  March,
  April,
  May,
  June,
  July,
  August,
  September,
  October,
  November,
  December,
}

extension on Months {
  String string() {
    return toString().split('.').last;
  }
}

class CalendarCarousel extends StatefulWidget {
  /// Specifies the spacing between the month and the weekdays
  final EdgeInsets? monthSelectorMargin;
  final EdgeInsets? weekdayMargin;
  final Function(DateTime) onChanged;
  final WeekdayAbbreviation weekdayAbb;
  final Color? selectedColor,
      unselectedColor,
      selectedTextColor,
      unselectedTextColor;
  final double? dayCarouselHeight;
  final DateTime? firstDate, lastDate;
  final bool showYearAlways;

  const CalendarCarousel({
    Key? key,
    required this.onChanged,
    this.monthSelectorMargin,
    this.weekdayMargin,
    this.weekdayAbb = WeekdayAbbreviation.three,
    this.selectedColor,
    this.unselectedColor,
    this.selectedTextColor,
    this.unselectedTextColor,
    this.dayCarouselHeight,
    this.firstDate,
    this.lastDate,
    this.showYearAlways = false,
  }) : super(key: key);

  @override
  State<CalendarCarousel> createState() => _CalendarCarouselState();
}

class _CalendarCarouselState extends State<CalendarCarousel> {
  /// The Value here should be the initial page index of the carousel
  int selectedDay = 10; // cannot be zero
  int selectedMonth = 12; // ranges between 1 - 12
  int? selectedYear;

  final CarouselController monthCarouselController = CarouselController();
  final CarouselController dayCarouselController = CarouselController();

  @override
  void initState() {
    super.initState();

    if (mounted) {
      setState(() {
        selectedDay = _currentDate.day;
        selectedMonth = _currentDate.month;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: const EdgeInsets.symmetric(vertical: 8.0),
      child: Column(
        children: [
          Container(
            margin: widget.monthSelectorMargin,
            width: MediaQuery.of(context).size.width,
            child: Stack(
              alignment: Alignment.center,
              children: [
                Row(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    Expanded(
                      child: InkWell(
                        onTap: () {
                          selectDate(context,
                                  firstDate: widget.firstDate,
                                  lastDate: widget.lastDate)
                              .then((value) {
                            setState(() {
                              selectedYear = selectedYear != _currentDate.year
                                  ? value?.year
                                  : null;
                              selectedMonth = value?.month ?? 1;
                              selectedDay = value?.day ?? 1;

                              monthCarouselController
                                  .animateToPage(selectedMonth - 1);
                              dayCarouselController
                                  .animateToPage(selectedDay - 1);
                            });
                          });
                        },
                        child: CarouselSlider.builder(
                            carouselController: monthCarouselController,
                            itemCount: selectedYear == _currentDate.year ||
                                    selectedYear == null
                                ? _currentDate.month
                                : Months.values.length,
                            itemBuilder: (context, index, realIndex) {
                              return Padding(
                                padding:
                                    const EdgeInsets.symmetric(vertical: 8.0),
                                child: Text(
                                  Months.values[index].string() +
                                      (selectedYear != null
                                          ? " $selectedYear"
                                          : (widget.showYearAlways
                                              ? ' ${DateTime.now().year}'
                                              : '')),
                                  textAlign: TextAlign.center,
                                  style: const TextStyle(
                                    color: Colors.black,
                                    fontWeight: FontWeight.w600,
                                    fontSize: 16,
                                  ),
                                ),
                              );
                            },
                            options: CarouselOptions(
                                height: 40,
                                enlargeCenterPage: true,
                                enableInfiniteScroll: false,
                                initialPage: selectedMonth - 1,
                                enlargeStrategy:
                                    CenterPageEnlargeStrategy.height,
                                viewportFraction: 0.8,
                                onPageChanged: (index, reason) {
                                  setState(() {
                                    selectedMonth = index + 1;
                                  });
                                  // widget.onChanged(currentDate);
                                })),
                      ),
                    ),
                  ],
                ),
                Positioned(
                    left: 0.0,
                    child: IconButton(
                      onPressed: () {
                        monthCarouselController.previousPage();
                      },
                      icon: const Icon(
                        Icons.chevron_left,
                        size: 32,
                      ),
                    )),
                Positioned(
                    right: 0.0,
                    child: IconButton(
                      onPressed: () {
                        monthCarouselController.nextPage();
                      },
                      icon: const Icon(
                        Icons.chevron_right,
                        size: 32,
                      ),
                    )),
              ],
            ),
          ),
          SizedBox(
            height: widget.dayCarouselHeight ?? 70,
            width: MediaQuery.of(context).size.width,
            child: CarouselSlider.builder(
              itemCount: month(
                      month: Months.values[selectedMonth - 1],
                      year: selectedYear)
                  .length,
              itemBuilder: (context, index, realIndex) {
                bool isCurrent = (selectedDay - 1) == realIndex;
                return Wrap(
                  children: [
                    Container(
                      constraints:
                          const BoxConstraints(minHeight: 60, minWidth: 40),
                      padding: const EdgeInsets.all(4.0),
                      decoration: BoxDecoration(
                        borderRadius: BorderRadius.circular(8),
                        color: isCurrent
                            ? widget.selectedColor ??
                                Theme.of(context).primaryColor
                            : widget.unselectedColor ?? Colors.transparent,
                      ),
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          Padding(
                            padding: widget.weekdayMargin ?? EdgeInsets.zero,
                            child: Text(
                              month(
                                      month: Months.values[selectedMonth - 1],
                                      year: selectedYear)
                                  .values
                                  .elementAt(index)
                                  .substring(0, widget.weekdayAbb.length),
                              style: TextStyle(
                                color: isCurrent
                                    ? widget.selectedTextColor ?? Colors.white
                                    : widget.unselectedTextColor ??
                                        const Color(0xFFBCC1CD),
                                fontWeight: FontWeight.w600,
                                fontSize: 14,
                              ),
                            ),
                          ),
                          Padding(
                            padding: const EdgeInsets.all(4.0),
                            child: Text(
                              month(
                                      month: Months.values[selectedMonth - 1],
                                      year: selectedYear)
                                  .keys
                                  .elementAt(index),
                              style: TextStyle(
                                color: isCurrent ? Colors.white : Colors.black,
                                fontWeight: FontWeight.w600,
                                fontSize: 16,
                              ),
                            ),
                          ),
                        ],
                      ),
                    ),
                  ],
                );
              },
              carouselController: dayCarouselController,
              options: CarouselOptions(
                  enlargeCenterPage: true,
                  enableInfiniteScroll: false,
                  initialPage: selectedDay - 1,
                  aspectRatio: 9 / 16,
                  viewportFraction: 0.15,
                  enlargeStrategy: CenterPageEnlargeStrategy.height,
                  onPageChanged: (index, reason) {
                    setState(() {
                      selectedDay = index + 1;
                    });
                    widget.onChanged(DateTime(selectedYear ?? _currentDate.year,
                        selectedMonth, selectedDay));
                  }),
              // items: imageSliders,
            ),
          )
        ],
      ),
    );
  }
}

Future<DateTime?> selectDate(BuildContext context,
    {DateTime? firstDate, DateTime? lastDate}) async {
  // final DateTime? picked =
  return await showDatePicker(
    context: context,
    initialDate: DateTime.now(),
    initialDatePickerMode: DatePickerMode.day,
    firstDate: firstDate ?? DateTime(2020),
    lastDate: lastDate ?? DateTime.now(),
  );
}

enum WeekdayAbbreviation {
  one,
  two,
  three,
  full,
}

extension on WeekdayAbbreviation {
  int? get length {
    switch (this) {
      case WeekdayAbbreviation.one:
        return 1;
      case WeekdayAbbreviation.two:
        return 2;
      case WeekdayAbbreviation.three:
        return 3;
      case WeekdayAbbreviation.full:
        return null;
    }
  }
}


lib/src/carousel_calendar.dart
import 'package:carousel_slider/carousel_slider.dart';
import 'package:flutter/material.dart';

final DateTime _currentDate = DateTime.now();

Map<String, String> month({required Months month, int? year}) {
  Map<String, String> map = {};
  var days = 31;
  switch (month) {
    case Months.february:
      days = 28;
      break;
    case Months.april:
    case Months.june:
    case Months.september:
    case Months.november:
      days = 30;
      break;
    default:
      days = 31;
      break;
  }
  year = year ?? _currentDate.year;

  if ((month.index + 1) == _currentDate.month && year == _currentDate.year) {
    days = _currentDate.day;
  }

  for (var day in List.generate(days, (index) => index + 1)) {
    map.addAll(
        {day.toString(): getDay(DateTime(year, month.index + 1, day).weekday)});
  }
  return map;
}

String getDay(int element) {
  switch (element % 7) {
    case 1:
      return "Monday";
    case 2:
      return "Tuesday";
    case 3:
      return "Wednesday";
    case 4:
      return "Thursday";
    case 5:
      return "Friday";
    case 6:
      return "Saturday";
    default:
      return "Sunday";
  }
}

enum Months {
  january,
  february,
  march,
  april,
  may,
  june,
  july,
  august,
  september,
  october,
  november,
  december,
}

extension on Months {
  String string() {
    return toString().split('.').last;
  }

  String capitalized() {
    List<String> separate = string().split(RegExp(r"(?=[A-Z])"));
    String labelMerge = "";

    for (var la in separate) {
      labelMerge += "${la.capitalize()} ";
    }

    return labelMerge.trimRight();
  }
}

class CalendarCarousel extends StatefulWidget {
  /// Specifies the spacing between the month and the weekdays
  final EdgeInsets? monthSelectorMargin;
  final EdgeInsets? weekdayMargin;
  final Function(DateTime) onChanged;
  final WeekdayAbbreviation weekdayAbb;
  final Color? selectedColor,
      unselectedColor,
      selectedTextColor,
      unselectedTextColor;
  final double? dayCarouselHeight;
  final DateTime? firstDate, lastDate;
  final bool showYearAlways;
  final DateTime? value;

  const CalendarCarousel({
    Key? key,
    required this.onChanged,
    this.monthSelectorMargin,
    this.weekdayMargin,
    this.weekdayAbb = WeekdayAbbreviation.three,
    this.selectedColor,
    this.unselectedColor,
    this.selectedTextColor,
    this.unselectedTextColor,
    this.dayCarouselHeight,
    this.firstDate,
    this.lastDate,
    this.showYearAlways = false,
    this.value,
  }) : super(key: key);

  @override
  State<CalendarCarousel> createState() => _CalendarCarouselState();
}

class _CalendarCarouselState extends State<CalendarCarousel> {
  /// The Value here should be the initial page index of the carousel
  int selectedDay = 10; // cannot be zero
  int selectedMonth = 12; // ranges between 1 - 12
  int? selectedYear;

  final CarouselController monthCarouselController = CarouselController();
  final CarouselController dayCarouselController = CarouselController();

  @override
  void initState() {
    super.initState();

    if (mounted) {
      setState(() {
        selectedDay = widget.value?.day ?? _currentDate.day;
        selectedMonth = widget.value?.month ?? _currentDate.month;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: const EdgeInsets.symmetric(vertical: 8.0),
      child: Column(
        children: [
          Container(
            margin: widget.monthSelectorMargin,
            width: MediaQuery.of(context).size.width,
            child: Stack(
              alignment: Alignment.center,
              children: [
                Row(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    Expanded(
                      child: InkWell(
                        onTap: () {
                          selectDate(context,
                                  firstDate: widget.firstDate,
                                  lastDate: widget.lastDate)
                              .then((value) {
                            setState(() {
                              selectedYear = selectedYear != _currentDate.year
                                  ? value?.year
                                  : null;
                              selectedMonth = value?.month ?? 1;
                              selectedDay = value?.day ?? 1;

                              monthCarouselController
                                  .animateToPage(selectedMonth - 1);
                              dayCarouselController
                                  .animateToPage(selectedDay - 1);
                            });
                          });
                        },
                        child: CarouselSlider.builder(
                            carouselController: monthCarouselController,
                            itemCount: selectedYear == _currentDate.year ||
                                    selectedYear == null
                                ? _currentDate.month
                                : Months.values.length,
                            itemBuilder: (context, index, realIndex) {
                              return Padding(
                                padding:
                                    const EdgeInsets.symmetric(vertical: 8.0),
                                child: Text(
                                  Months.values[index].capitalized() +
                                      (selectedYear != null
                                          ? " $selectedYear"
                                          : (widget.showYearAlways
                                              ? ' ${DateTime.now().year}'
                                              : '')),
                                  textAlign: TextAlign.center,
                                  style: const TextStyle(
                                    color: Colors.black,
                                    fontWeight: FontWeight.w600,
                                    fontSize: 16,
                                  ),
                                ),
                              );
                            },
                            options: CarouselOptions(
                                height: 40,
                                enlargeCenterPage: true,
                                enableInfiniteScroll: false,
                                initialPage: selectedMonth - 1,
                                enlargeStrategy:
                                    CenterPageEnlargeStrategy.height,
                                viewportFraction: 0.8,
                                onPageChanged: (index, reason) {
                                  setState(() {
                                    selectedMonth = index + 1;
                                  });
                                  // widget.onChanged(currentDate);
                                })),
                      ),
                    ),
                  ],
                ),
                Positioned(
                    left: 0.0,
                    child: IconButton(
                      onPressed: () {
                        monthCarouselController.previousPage();
                      },
                      icon: const Icon(
                        Icons.chevron_left,
                        size: 32,
                      ),
                    )),
                Positioned(
                    right: 0.0,
                    child: IconButton(
                      onPressed: () {
                        monthCarouselController.nextPage();
                      },
                      icon: const Icon(
                        Icons.chevron_right,
                        size: 32,
                      ),
                    )),
              ],
            ),
          ),
          SizedBox(
            height: widget.dayCarouselHeight ?? 70,
            width: MediaQuery.of(context).size.width,
            child: CarouselSlider.builder(
              itemCount: month(
                      month: Months.values[selectedMonth - 1],
                      year: selectedYear)
                  .length,
              itemBuilder: (context, index, realIndex) {
                bool isCurrent = (selectedDay - 1) == realIndex;
                return Wrap(
                  children: [
                    Container(
                      constraints:
                          const BoxConstraints(minHeight: 60, minWidth: 40),
                      padding: const EdgeInsets.all(4.0),
                      decoration: BoxDecoration(
                        borderRadius: BorderRadius.circular(8),
                        color: isCurrent
                            ? widget.selectedColor ??
                                Theme.of(context).primaryColor
                            : widget.unselectedColor ?? Colors.transparent,
                      ),
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          Padding(
                            padding: widget.weekdayMargin ?? EdgeInsets.zero,
                            child: Text(
                              month(
                                      month: Months.values[selectedMonth - 1],
                                      year: selectedYear)
                                  .values
                                  .elementAt(index)
                                  .substring(0, widget.weekdayAbb.length),
                              style: TextStyle(
                                color: isCurrent
                                    ? widget.selectedTextColor ?? Colors.white
                                    : widget.unselectedTextColor ??
                                        const Color(0xFFBCC1CD),
                                fontWeight: FontWeight.w600,
                                fontSize: 14,
                              ),
                            ),
                          ),
                          Padding(
                            padding: const EdgeInsets.all(4.0),
                            child: Text(
                              month(
                                      month: Months.values[selectedMonth - 1],
                                      year: selectedYear)
                                  .keys
                                  .elementAt(index),
                              style: TextStyle(
                                color: isCurrent ? Colors.white : Colors.black,
                                fontWeight: FontWeight.w600,
                                fontSize: 16,
                              ),
                            ),
                          ),
                        ],
                      ),
                    ),
                  ],
                );
              },
              carouselController: dayCarouselController,
              options: CarouselOptions(
                  enlargeCenterPage: true,
                  enableInfiniteScroll: false,
                  initialPage: selectedDay - 1,
                  aspectRatio: 9 / 16,
                  viewportFraction: 0.15,
                  enlargeStrategy: CenterPageEnlargeStrategy.height,
                  onPageChanged: (index, reason) {
                    setState(() {
                      selectedDay = index + 1;
                    });
                    widget.onChanged(DateTime(selectedYear ?? _currentDate.year,
                        selectedMonth, selectedDay));
                  }),
              // items: imageSliders,
            ),
          )
        ],
      ),
    );
  }
}

Future<DateTime?> selectDate(BuildContext context,
    {DateTime? firstDate, DateTime? lastDate}) async {
  // final DateTime? picked =
  return await showDatePicker(
    context: context,
    initialDate: DateTime.now(),
    initialDatePickerMode: DatePickerMode.day,
    firstDate: firstDate ?? DateTime(2020),
    lastDate: lastDate ?? DateTime.now(),
  );
}

enum WeekdayAbbreviation {
  one,
  two,
  three,
  full,
}

extension on WeekdayAbbreviation {
  int? get length {
    switch (this) {
      case WeekdayAbbreviation.one:
        return 1;
      case WeekdayAbbreviation.two:
        return 2;
      case WeekdayAbbreviation.three:
        return 3;
      case WeekdayAbbreviation.full:
        return null;
    }
  }
}

extension StringCasingExtension on String {
  String capitalize() =>
      length > 0 ? '${this[0].toUpperCase()}${substring(1).toLowerCase()}' : '';
}


example/lib/main.dart
import 'package:carousel_calendar/carousel_calendar.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Calendar Carousel'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  DateTime? date;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          children: [
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: CalendarCarousel(
                monthSelectorMargin: EdgeInsets.only(bottom: 2.0),
                weekdayMargin: EdgeInsets.symmetric(vertical: 12.0),
                weekdayAbb: WeekdayAbbreviation.two,
                dayCarouselHeight: 120,
                showYearAlways: true,
                onChanged: (nm) {
                  setState(() {
                    date = nm;
                  });
                },
              ),
            ),
            Expanded(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  const Text(
                    'Selected Date',
                  ),
                  Text(
                    '${date != null ? "${date!.day}-${date!.month}-${date!.year}" : "No selection yet."}',
                    style: Theme.of(context).textTheme.headlineMedium,
                  ),
                  Visibility(
                    visible: false,
                    child: const Text(
                      'You have pushed the button this many times:',
                    ),
                  ),
                  Visibility(
                    visible: false,
                    child: Text(
                      '$_counter',
                      style: Theme.of(context).textTheme.headlineMedium,
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: Visibility(
        visible: false,
        child: FloatingActionButton(
          onPressed: _incrementCounter,
          tooltip: 'Increment',
          child: const Icon(Icons.add),
        ),
      ),
    );
  }
}
